File: relationship-trends/js/parsers.js
// Parsers for communication exports.
//
// Supported inputs:
//  - calls.xml / sms.xml from the Android "SMS Backup & Restore" app
//  - a generic CSV (see README for the column schema)
//
// Everything parses into a flat list of Event objects:
//   { contactName, number, kind: 'text'|'call', direction: 'in'|'out'|'missed',
//     ts (epoch ms), durationSec }

const XML_ENTITIES = {
  '&amp;': '&',
  '&lt;': '<',
  '&gt;': '>',
  '&quot;': '"',
  '&apos;': "'",
};

export function decodeEntities(str) {
  return str
    .replace(/&#(\d+);/g, (_, n) => String.fromCodePoint(Number(n)))
    .replace(/&#x([0-9a-fA-F]+);/g, (_, n) => String.fromCodePoint(parseInt(n, 16)))
    .replace(/&(amp|lt|gt|quot|apos);/g, (m) => XML_ENTITIES[m]);
}

// Scan self-closing/opening tags of a given name and return attribute maps.
// SMS Backup & Restore output is flat, machine-generated XML, so a tag scanner
// is sufficient and works identically in the browser and in Node tests.
export function scanTags(xml, tagName) {
  const results = [];
  const tagRe = new RegExp(`<${tagName}\\b([^>]*?)/?>`, 'g');
  const attrRe = /([\w:]+)\s*=\s*"([^"]*)"/g;
  let m;
  while ((m = tagRe.exec(xml)) !== null) {
    const attrs = {};
    let a;
    attrRe.lastIndex = 0;
    while ((a = attrRe.exec(m[1])) !== null) {
      attrs[a[1]] = decodeEntities(a[2]);
    }
    results.push(attrs);
  }
  return results;
}

function cleanName(name) {
  if (!name || name === '(Unknown)' || name === 'null') return '';
  return name.trim();
}

// calls.xml: <call number=".." duration=".." date=".." type=".." contact_name=".." />
// type: 1 incoming, 2 outgoing, 3 missed, 5 rejected (treated as missed)
export function parseCallsXml(xml) {
  return scanTags(xml, 'call')
    .map((c) => {
      const type = Number(c.type);
      const direction = type === 2 ? 'out' : type === 1 ? 'in' : 'missed';
      return {
        contactName: cleanName(c.contact_name),
        number: c.number || '',
        kind: 'call',
        direction,
        ts: Number(c.date),
        durationSec: Number(c.duration) || 0,
      };
    })
    .filter((e) => Number.isFinite(e.ts) && e.ts > 0);
}

// sms.xml: <sms address=".." date=".." type=".." contact_name=".." />
// type: 1 received, 2 sent (other types — drafts, queued — are skipped)
export function parseSmsXml(xml) {
  return scanTags(xml, 'sms')
    .map((s) => {
      const type = Number(s.type);
      if (type !== 1 && type !== 2) return null;
      return {
        contactName: cleanName(s.contact_name),
        number: s.address || '',
        kind: 'text',
        direction: type === 2 ? 'out' : 'in',
        ts: Number(s.date),
        durationSec: 0,
      };
    })
    .filter((e) => e && Number.isFinite(e.ts) && e.ts > 0);
}

// Generic CSV with a header row. Recognized columns (case-insensitive):
//   date (ISO 8601 or epoch ms), kind (text|call|meet),
//   direction (in|out|missed; blank for meet), contact, number,
//   duration_seconds
export function parseCsv(text) {
  const rows = splitCsv(text);
  if (rows.length < 2) return [];
  const header = rows[0].map((h) => h.trim().toLowerCase());
  const col = (name) => header.indexOf(name);
  const iDate = col('date');
  const iKind = col('kind');
  const iDir = col('direction');
  const iContact = col('contact');
  const iNumber = col('number');
  const iDur = col('duration_seconds');
  if (iDate === -1 || iKind === -1 || iDir === -1) {
    throw new Error('CSV must have at least "date", "kind" and "direction" columns');
  }
  const events = [];
  for (let r = 1; r < rows.length; r++) {
    const row = rows[r];
    if (row.length === 1 && row[0].trim() === '') continue;
    const rawDate = (row[iDate] || '').trim();
    const ts = /^\d+$/.test(rawDate) ? Number(rawDate) : Date.parse(rawDate);
    const kind = (row[iKind] || '').trim().toLowerCase();
    let direction = (row[iDir] || '').trim().toLowerCase();
    if (!Number.isFinite(ts)) continue;
    if (kind === 'meet') direction = 'met'; // an in-person meet has no direction
    else if (kind !== 'text' && kind !== 'call') continue;
    else if (!['in', 'out', 'missed'].includes(direction)) continue;
    events.push({
      contactName: iContact === -1 ? '' : cleanName(row[iContact]),
      number: iNumber === -1 ? '' : (row[iNumber] || '').trim(),
      kind,
      direction,
      ts,
      durationSec: iDur === -1 ? 0 : Number(row[iDur]) || 0,
    });
  }
  return events;
}

function splitCsv(text) {
  const rows = [];
  let row = [];
  let field = '';
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') { field += '"'; i++; }
        else inQuotes = false;
      } else field += ch;
    } else if (ch === '"') {
      inQuotes = true;
    } else if (ch === ',') {
      row.push(field); field = '';
    } else if (ch === '\n' || ch === '\r') {
      if (ch === '\r' && text[i + 1] === '\n') i++;
      row.push(field); field = '';
      rows.push(row); row = [];
    } else {
      field += ch;
    }
  }
  if (field !== '' || row.length > 0) { row.push(field); rows.push(row); }
  return rows;
}

// Sniff the format of a file's text content and parse accordingly.
export function parseAny(text) {
  const head = text.slice(0, 2000);
  if (/<calls\b/.test(head) || /<call\b/.test(head)) return parseCallsXml(text);
  if (/<smses\b/.test(head) || /<sms\b/.test(head)) return parseSmsXml(text);
  return parseCsv(text);
}

File: relationship-trends/ios/KindredCore/Sources/KindredCore/Parsers.swift
import Foundation

/// Parsers for communication exports — the Kindred CSV interchange format
/// (also produced by export-tools/export_comms.py) and the Android
/// "SMS Backup & Restore" calls.xml / sms.xml files.
public enum Parsers {
    public enum ParseError: LocalizedError {
        case missingColumns

        public var errorDescription: String? {
            switch self {
            case .missingColumns:
                return "CSV must have at least \"date\", \"kind\" and \"direction\" columns"
            }
        }
    }

    /// Sniff the format of a file's text content and parse accordingly.
    public static func parseAny(_ text: String) throws -> [CommEvent] {
        let head = String(text.prefix(2000))
        if head.contains("<calls") || head.contains("<call ") { return parseCallsXml(text) }
        if head.contains("<smses") || head.contains("<sms ") { return parseSmsXml(text) }
        return try parseCsv(text)
    }

    // MARK: - CSV

    /// Columns (case-insensitive): date (ISO 8601 or epoch ms), kind
    /// (text|call|meet), direction (in|out|missed; blank for meet), contact,
    /// number, duration_seconds. Extra columns are ignored.
    public static func parseCsv(_ text: String) throws -> [CommEvent] {
        let rows = splitCsv(text)
        guard rows.count >= 2 else { return [] }
        let header = rows[0].map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
        guard
            let iDate = header.firstIndex(of: "date"),
            let iKind = header.firstIndex(of: "kind"),
            let iDir = header.firstIndex(of: "direction")
        else { throw ParseError.missingColumns }
        let iContact = header.firstIndex(of: "contact")
        let iNumber = header.firstIndex(of: "number")
        let iDur = header.firstIndex(of: "duration_seconds")

        var events: [CommEvent] = []
        for row in rows.dropFirst() {
            if row.count == 1 && row[0].trimmingCharacters(in: .whitespaces).isEmpty { continue }
            func field(_ i: Int?) -> String {
                guard let i, i < row.count else { return "" }
                return row[i].trimmingCharacters(in: .whitespaces)
            }
            guard let ts = parseDateMs(field(iDate)) else { continue }
            guard let kind = EventKind(rawValue: field(iKind).lowercased()) else { continue }
            let direction: Direction
            if kind == .meet {
                direction = .met // an in-person meet has no direction
            } else {
                guard let d = Direction(rawValue: field(iDir).lowercased()), d != .met else { continue }
                direction = d
            }
            events.append(CommEvent(
                contactName: cleanName(field(iContact)),
                number: field(iNumber),
                kind: kind,
                direction: direction,
                timestampMs: ts,
                durationSec: Double(field(iDur)) ?? 0
            ))
        }
        return events
    }

    static func parseDateMs(_ raw: String) -> Double? {
        guard !raw.isEmpty else { return nil }
        if raw.allSatisfy(\.isNumber) { return Double(raw) }
        let iso = ISO8601DateFormatter()
        if let d = iso.date(from: raw) { return d.timeIntervalSince1970 * 1000 }
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = iso.date(from: raw) { return d.timeIntervalSince1970 * 1000 }
        let plain = DateFormatter()
        plain.locale = Locale(identifier: "en_US_POSIX")
        plain.timeZone = TimeZone(identifier: "UTC")
        plain.dateFormat = "yyyy-MM-dd"
        if let d = plain.date(from: raw) { return d.timeIntervalSince1970 * 1000 }
        return nil
    }

    static func splitCsv(_ text: String) -> [[String]] {
        var rows: [[String]] = []
        var row: [String] = []
        var field = ""
        var inQuotes = false
        var i = text.startIndex
        while i < text.endIndex {
            let ch = text[i]
            if inQuotes {
                if ch == "\"" {
                    let next = text.index(after: i)
                    if next < text.endIndex && text[next] == "\"" {
                        field.append("\"")
                        i = next
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(ch)
                }
            } else if ch == "\"" {
                inQuotes = true
            } else if ch == "," {
                row.append(field)
                field = ""
            } else if ch == "\n" || ch == "\r" {
                if ch == "\r" {
                    let next = text.index(after: i)
                    if next < text.endIndex && text[next] == "\n" { i = next }
                }
                row.append(field)
                field = ""
                rows.append(row)
                row = []
            } else {
                field.append(ch)
            }
        }
        if !field.isEmpty || !row.isEmpty {
            row.append(field)
            rows.append(row)
        }
        return rows
    }

    // MARK: - SMS Backup & Restore XML

    /// calls.xml — type: 1 incoming, 2 outgoing, 3 missed, 5 rejected.
    public static func parseCallsXml(_ xml: String) -> [CommEvent] {
        scanTags(xml, tagName: "call").compactMap { attrs in
            guard let ts = Double(attrs["date"] ?? ""), ts > 0 else { return nil }
            let direction: Direction
            switch attrs["type"] {
            case "2": direction = .outgoing
            case "1": direction = .incoming
            default: direction = .missed
            }
            return CommEvent(
                contactName: cleanName(attrs["contact_name"] ?? ""),
                number: attrs["number"] ?? "",
                kind: .call,
                direction: direction,
                timestampMs: ts,
                durationSec: direction == .missed ? 0 : Double(attrs["duration"] ?? "") ?? 0
            )
        }
    }

    /// sms.xml — type: 1 received, 2 sent (drafts/queued are skipped).
    public static func parseSmsXml(_ xml: String) -> [CommEvent] {
        scanTags(xml, tagName: "sms").compactMap { attrs in
            guard let ts = Double(attrs["date"] ?? ""), ts > 0 else { return nil }
            let direction: Direction
            switch attrs["type"] {
            case "1": direction = .incoming
            case "2": direction = .outgoing
            default: return nil
            }
            return CommEvent(
                contactName: cleanName(attrs["contact_name"] ?? ""),
                number: attrs["address"] ?? "",
                kind: .text,
                direction: direction,
                timestampMs: ts
            )
        }
    }

    /// Scan flat, machine-generated XML for a tag's attribute maps.
    static func scanTags(_ xml: String, tagName: String) -> [[String: String]] {
        guard
            let tagRe = try? NSRegularExpression(pattern: "<\(tagName)\\b([^>]*?)/?>"),
            let attrRe = try? NSRegularExpression(pattern: "([\\w:]+)\\s*=\\s*\"([^\"]*)\"")
        else { return [] }
        let ns = xml as NSString
        return tagRe.matches(in: xml, range: NSRange(location: 0, length: ns.length)).map { m in
            let body = ns.substring(with: m.range(at: 1))
            let bodyNs = body as NSString
            var attrs: [String: String] = [:]
            for a in attrRe.matches(in: body, range: NSRange(location: 0, length: bodyNs.length)) {
                let key = bodyNs.substring(with: a.range(at: 1))
                attrs[key] = decodeEntities(bodyNs.substring(with: a.range(at: 2)))
            }
            return attrs
        }
    }

    static func decodeEntities(_ str: String) -> String {
        guard str.contains("&") else { return str }
        var out = str
        for (entity, ch) in [("&lt;", "<"), ("&gt;", ">"), ("&quot;", "\""), ("&apos;", "'")] {
            out = out.replacingOccurrences(of: entity, with: ch)
        }
        // Numeric entities: &#65; and &#x42;
        if let re = try? NSRegularExpression(pattern: "&#(x?)([0-9a-fA-F]+);") {
            let ns = out as NSString
            var result = ""
            var last = 0
            for m in re.matches(in: out, range: NSRange(location: 0, length: ns.length)) {
                result += ns.substring(with: NSRange(location: last, length: m.range.location - last))
                let isHex = m.range(at: 1).length > 0
                let digits = ns.substring(with: m.range(at: 2))
                if let code = UInt32(digits, radix: isHex ? 16 : 10), let scalar = Unicode.Scalar(code) {
                    result.append(Character(scalar))
                } else {
                    result += ns.substring(with: m.range)
                }
                last = m.range.location + m.range.length
            }
            result += ns.substring(from: last)
            out = result
        }
        return out.replacingOccurrences(of: "&amp;", with: "&")
    }

    static func cleanName(_ name: String) -> String {
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        if trimmed == "(Unknown)" || trimmed == "null" { return "" }
        return trimmed
    }
}
